crates/signature-mpc/src/twopc_mpc_protocols/mod.rs

```diff
@@ -499,85 +477,89 @@ fn decrypt_single_signature(
+/// Page 27 Optimized Threshold Decryption.
+pub fn decrypt_signatures_decentralized_party_sign(
     decryption_key_share_public_parameters: DecryptionPublicParameters,
-    decryption_shares: HashMap<
+    // PartyID => (partial_signature_decryption_share,
+    // masked_nonce_decryption_share) per Message.
+    batch_decryption_shares: HashMap<
         PartyID,
+        // Vec<(partial_signature_decryption_share, masked_nonce_decryption_share)>.
         Vec<(PaillierModulusSizedNumber, PaillierModulusSizedNumber)>,
     >,
     public_nonce_encrypted_partial_signature_and_proofs: Vec<
         PublicNonceEncryptedPartialSignatureAndProof<ProtocolContext>,
     >,
     signature_threshold_decryption_round_parties: Vec<SignatureThresholdDecryptionParty>,
 ) -> std::result::Result<Vec<Vec<u8>>, DecryptionError> {
-    let decrypters: Vec<PartyID> = decryption_shares
+    let threshold = decryption_key_share_public_parameters.threshold as usize;
+    let threshold_decrypters: Vec<PartyID> = batch_decryption_shares
         .keys()
-        .take(decryption_key_share_public_parameters.threshold.into())
+        .take(threshold)
         .copied()
         .collect();
 
-    // todo(itay): explain this, why public_nonce_encrypted_partial_signature_and_proofs,
-    // todo: why not decryption_shares?
-    // todo(zeev): fix.
-    let decryption_shares: Vec<(HashMap<_, _>, HashMap<_, _>)> = (0
-        ..public_nonce_encrypted_partial_signature_and_proofs.len())
+    let decryption_shares: Vec<(
+        HashMap<PartyID, PaillierModulusSizedNumber>,
+        HashMap<PartyID, PaillierModulusSizedNumber>,
+    )> = (0..public_nonce_encrypted_partial_signature_and_proofs.len())
         .map(|i| {
-            decryption_shares
+            batch_decryption_shares
                 .iter()
-                .filter(|(party_id, _)| decrypters.contains(party_id))
-                .map(|(party_id, decryption_share)| {
-                    let (partial_signature_decryption_shares, masked_nonce_decryption_shares) =
-                        decryption_share[i];
-                    (
-                        (*party_id, partial_signature_decryption_shares),
-                        (*party_id, masked_nonce_decryption_shares),
-                    )
-                })
-                .unzip()
+                .filter(|(party_id, _)| threshold_decrypters.contains(party_id))
+                .fold(
+                    (HashMap::new(), HashMap::new()),
+                    |(mut partial_map, mut masked_map), (party_id, decryption_shares)| {
+                        let (partial_signature_decryption_shares, masked_nonce_decryption_shares) =
+                            decryption_shares[i];
+                        partial_map.insert(*party_id, partial_signature_decryption_shares);
+                        masked_map.insert(*party_id, masked_nonce_decryption_shares);
+                        (partial_map, masked_map)
+                    },
+                )
```

Comment:
why did you change this `fold` with mutable elements rather the old `map`? I believe this code is even less readable, and I still didn't understand what the purpose of this code is (my question to omer remained unanswered.)